Reorganize relocation table (#224)
riscv-non-isa/riscv-elf-psabi-doc: riscv-elf.adoc

machine word sized relocations that are used for symbol addresses in the
Global Offset Table or DWARF meta data.

> <<reloc-table>> provides details of the RISC-V ELF relocations, means of each
> columns are list below:

Reviewer: ```suggestion
<<reloc-table>> provides details of the RISC-V ELF relocations; the meaning of each
column is given below:
```

> Enum:: The enum value of the relocation, used in the r_info field of
> the ELF32_R_TYPE/ELF64_R_TYPE.

Reviewer: ```suggestion
Enum:: The number of the relocation, encoded in the r_info field
```

> ELF Reloc Type:: The mnemonic name for the relocation.

Reviewer: ```suggestion
ELF Reloc Type:: The name of the relocation
```

> Type:: Indicate this relocation is static or runtime relocation type:

Reviewer: ```suggestion
Type:: Whether the relocation is a static or runtime relocation:
```

+

> - Static relocations are must resolved by static linker.

Reviewer: ```suggestion
- Static relocations are always resolved by the static linker
```

> - Runtime relocations are resolved by static linker or dynamic linker.

Reviewer: ```suggestion
- Runtime relocations can be resolved by both static and dynamic linkers
```

> +

Reviewer: ```suggestion
```

> Field::  Describe this relocation is apply to which field of data or
>          instruction, more detail see <<Field Symbols>>.

Reviewer: ```suggestion
Field:: Describes the set of bits affected by this relocation; see <<Field Symbols>> for the definitions of the individual types
```

> Caculation::  Formula for how to resolve the relocation value, definition of
>               symbol can be found <<Calculation Symbols>>

Reviewer: ```suggestion
Caculation:: Formula for how to resolve the relocation value; definitions of the
             symbols can be found in <<Calculation Symbols>>
```

> Description::  Description of this relocation.

Reviewer: ```suggestion
Description:: Additional information about the relocation
```

[[reloc-table]]
[cols=">2,7,3,6,11"]
[width=100%]
|===
| Enum          | ELF Reloc Type                         | Type    | Field / Calculation  | Description

.2+| 0       .2+| [.relocname]#R_RISCV_NONE#          .2+| None    |                   .2+|
                                                                  <|
.2+| 1       .2+| [.relocname]#R_RISCV_32#            .2+| Runtime | _word32_          .2+| 32-bit relocation
                                                                  <| S + A
.2+| 2       .2+| [.relocname]#R_RISCV_64#            .2+| Runtime | _word64_          .2+| 64-bit relocation
                                                                  <| S + A
.2+| 3       .2+| [.relocname]#R_RISCV_RELATIVE#      .2+| Runtime | _wordclass_       .2+| Relocation against a local symbol in a shared object
                                                                  <| B + A
.2+| 4       .2+| [.relocname]#R_RISCV_COPY#          .2+| Runtime |                   .2+| Must be in executable; not allowed in shared library
                                                                  <|

> .2+| 5       .2+| [.relocname]#R_RISCV_JUMP_SLOT#     .2+| Runtime | _wordclass_       .2+| Handled by PLT unless `LD_BIND_NOW`

Reviewer: LD_BIND_NOW is more complicated than that. I'd just say something like "Indicates the symbol associated with a PLT entry".

                                                                  <| S + A
.2+| 28      .2+| [.relocname]#R_RISCV_LO12_S#        .2+| Static  | _S-Type_          .2+| High 12 bits of 32-bit absolute address, `%lo(symbol)`
                                                                  <| S + A
.2+| 29      .2+| [.relocname]#R_RISCV_TPREL_HI20#    .2+| Static  | _U-Type_          .2+| High 20 bits of TLS LE thread pointer offset, `%tprel_hi(symbol)`
                                                                  <|
.2+| 30      .2+| [.relocname]#R_RISCV_TPREL_LO12_I#  .2+| Static  | _I-Type_          .2+| Low 12 bits of TLS LE thread pointer offset, `%tprel_lo(symbol)`
                                                                  <|
.2+| 31      .2+| [.relocname]#R_RISCV_TPREL_LO12_S#  .2+| Static  | _S-Type_          .2+| Low 12 bits of TLS LE thread pointer offset, `%tprel_lo(symbol)`
                                                                  <|
.2+| 32      .2+| [.relocname]#R_RISCV_TPREL_ADD#     .2+| Static  |                   .2+| TLS LE thread pointer usage, `%tprel_add(symbol)`
                                                                  <|

> .2+| 33      .2+| [.relocname]#R_RISCV_ADD8#          .2+| Static  | _word8_           .2+| 8-bit label addition .

Reviewer: Stray . at the end?

machine word sized relocations that are used for symbol addresses in the
Global Offset Table or DWARF meta data.

<<reloc-table>> provides details of the RISC-V ELF relocations; the meaning of each
column is given below:

Enum:: The number of the relocation, encoded in the r_info field

ELF Reloc Type:: The name of the relocation, omitted the prefix of `R_RISCV_` here.

Type:: Whether the relocation is a static or runtime relocation:
+
- Static relocations are always resolved by the static linker
- Runtime relocations can be resolved by both static and dynamic linkers

Field:: Describes the set of bits affected by this relocation; see <<Field Symbols>> for the definitions of the individual types

> Caculation:: Formula for how to resolve the relocation value; definitions of the
>              symbols can be found in <<Calculation Symbols>>

Reviewer: ```suggestion
Calculation:: Formula for how to resolve the relocation value; definitions of the
              symbols can be found in <<Calculation Symbols>>
```
